File: src/com/campusnet/model/NetworkInterface.java
package com.campusnet.model;

public class NetworkInterface {
    private final String name; // e.g., "eth0", "gi0/1"
    private final String macAddress;
    private String ipAddress;
    private String subnetMask;
    private boolean isUp;
    private NetworkLink connectedLink;
    private Device ownerDevice;

    public NetworkInterface(String name, String macAddress, String ipAddress, String subnetMask, Device ownerDevice) {
        this.name = name;
        this.macAddress = macAddress;
        this.ipAddress = ipAddress;
        this.subnetMask = subnetMask;
        this.isUp = true;
        this.ownerDevice = ownerDevice;
    }

    public NetworkInterface(String name, String macAddress, Device ownerDevice) {
        this(name, macAddress, null, null, ownerDevice);
    }

    public String getName() {
        return name;
    }

    public String getMacAddress() {
        return macAddress;
    }

    public String getIpAddress() {
        return ipAddress;
    }

    public void setIpAddress(String ipAddress) {
        this.ipAddress = ipAddress;
    }

    public String getSubnetMask() {
        return subnetMask;
    }

    public void setSubnetMask(String subnetMask) {
        this.subnetMask = subnetMask;
    }

    public boolean isUp() {
        return isUp && (connectedLink == null || connectedLink.isUp());
    }

    public void setUp(boolean up) {
        isUp = up;
    }

    public NetworkLink getConnectedLink() {
        return connectedLink;
    }

    public void setConnectedLink(NetworkLink connectedLink) {
        this.connectedLink = connectedLink;
    }

    public Device getOwnerDevice() {
        return ownerDevice;
    }

    public void setOwnerDevice(Device ownerDevice) {
        this.ownerDevice = ownerDevice;
    }

    public boolean isConfigured() {
        return ipAddress != null && !ipAddress.isEmpty();
    }

    public boolean isInSameSubnet(String otherIp) {
        if (ipAddress == null || subnetMask == null || otherIp == null) {
            return false;
        }
        try {
            long thisIpVal = ipToLong(ipAddress);
            long maskVal = ipToLong(subnetMask);
            long otherIpVal = ipToLong(otherIp);
            return (thisIpVal & maskVal) == (otherIpVal & maskVal);
        } catch (Exception e) {
            return false;
        }
    }

    public static long ipToLong(String ip) {
        String[] parts = ip.trim().split("\\.");
        long result = 0;
        for (int i = 0; i < 4; i++) {
            result = (result << 8) | (Integer.parseInt(parts[i]) & 0xFF);
        }
        return result;
    }

    public static String longToIp(long ip) {
        return String.format("%d.%d.%d.%d",
                (ip >> 24) & 0xFF,
                (ip >> 16) & 0xFF,
                (ip >> 8) & 0xFF,
                ip & 0xFF);
    }

    @Override
    public String toString() {
        return name + " (" + (ipAddress != null ? ipAddress : "No IP") + ", MAC: " + macAddress + ")";
    }
}

File: js/models/NetworkInterface.js
(function(exports) {
    class NetworkInterface {
        constructor(name, macAddress, ipAddress = null, subnetMask = null, ownerDevice = null) {
            this.name = name; // e.g., "eth0", "gi0/1"
            this.macAddress = macAddress;
            this.ipAddress = ipAddress;
            this.subnetMask = subnetMask;
            this.isUp = true;
            this.connectedLink = null;
            this.ownerDevice = ownerDevice;
        }

        static ipToLong(ip) {
            if (!ip) return 0;
            const parts = ip.trim().split('.');
            let res = 0;
            for (let i = 0; i < 4; i++) {
                res = ((res << 8) | (parseInt(parts[i], 10) & 0xFF)) >>> 0;
            }
            return res;
        }

        static longToIp(num) {
            return [
                (num >>> 24) & 0xFF,
                (num >>> 16) & 0xFF,
                (num >>> 8) & 0xFF,
                num & 0xFF
            ].join('.');
        }

        static getNetworkAddress(ip, mask) {
            if (!ip || !mask) return '0.0.0.0';
            const ipNum = NetworkInterface.ipToLong(ip);
            const maskNum = NetworkInterface.ipToLong(mask);
            return NetworkInterface.longToIp((ipNum & maskNum) >>> 0);
        }

        static getPrefixLength(mask) {
            if (!mask || mask === '0.0.0.0') return 0;
            let num = NetworkInterface.ipToLong(mask);
            let count = 0;
            while (num) {
                count += num & 1;
                num >>>= 1;
            }
            return count;
        }

        isInSameSubnet(otherIp) {
            if (!this.ipAddress || !this.subnetMask || !otherIp) return false;
            const thisNet = NetworkInterface.getNetworkAddress(this.ipAddress, this.subnetMask);
            const otherNet = NetworkInterface.getNetworkAddress(otherIp, this.subnetMask);
            return thisNet === otherNet;
        }

        isConfigured() {
            return !!this.ipAddress;
        }

        isActive() {
            return this.isUp && (!this.connectedLink || this.connectedLink.isUp);
        }
    }

    exports.NetworkInterface = NetworkInterface;
})(typeof module !== 'undefined' && module.exports ? module.exports : (window.CampusNet = window.CampusNet || {}));
